perfact/zodbsync/subcommand.py

```diff
@@ -113,6 +113,64 @@ def check_repo(self):
             if line.endswith(' HEAD')
         ][0].split()[0]
 
+    @staticmethod
+    def gitexec(func):
+        """
+        Decorator for wrapping an operation with playback:
+        - Stash unstaged changes away
+        - memorize the current commit
+        - do something
+        - check for conflicts
+        - play back changed objects (diff between old and new HEAD)
+        - unstash
+        """
+        @SubCommand.with_lock
+        def wrapper(self, *args, **kwargs):
+            # Check for unstaged changes
+            self.check_repo()
+
+            try:
+                func(self, *args, **kwargs)
+                # Are these all?
+                for fname in ['rebase-merge', 'rebase-apply',
+                              'CHERRY_PICK_HEAD', 'MERGE_HEAD', 'REVERT_HEAD']:
+                    path = os.path.join(self.sync.base_dir, '.git', fname)
+                    assert not os.path.exists(path), "Git state not clean"
+
+                files = {
+                    line for line in self.gitcmd_output(
+                        'diff', self.orig_commit, '--name-only'
+                    ).strip().split('\n')
+                    if line
+                }
+                conflicts = files & set(self.unstaged_changes)
+                assert not conflicts, "Change in unstaged files, aborting"
+
+                paths = sorted({
+                    filename[len(self.sync.site):].rsplit('/', 1)[0]
+                    for filename in files
+                })
```

Comment:
Wow, nice combos! Computing what these lines do make my nose bleed. Some comments would be helpful, especially `filename[len(self.sync.site):].rsplit('/', 1)[0]` is hard to read ( on sunday morning without coffee ).

Reply:
I concur. On first read, I don't know jack what's going on here.